server/src/test/java/io/crate/auth/JwtAuthenticationIntegrationTest.java

```diff
@@ -0,0 +1,222 @@
+                RSAPublicKey publicKey = (RSAPublicKey) keyFactory.generatePublic(publicKeySpec);
+
+                generator.writeStartObject();
+                generator.writeArrayFieldStart("keys");
+                generator.writeStartObject();
+                generator.writeStringField("e", BASE_64_URL_ENCODER.encodeToString(publicKey.getPublicExponent().toByteArray()));
+                generator.writeStringField("kid", KID);
+                generator.writeStringField("kty", "RSA");
+                generator.writeStringField("n", BASE_64_URL_ENCODER.encodeToString(publicKey.getModulus().toByteArray()));
+                generator.writeEndObject();
+                generator.writeEndArray();
+                generator.writeEndObject();
+                generator.close();
+            } catch (Exception e) {
+                throw new RuntimeException(e.getCause());
+            }
+        };
+    private HttpTestServer testServer;
+    private int port;
+    private RSAPrivateKey privateKey;
+
+    @Override
+    protected boolean addMockHttpTransport() {
+        return false;
+    }
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder()
+            .put(super.nodeSettings(nodeOrdinal))
+            .put("network.host", "127.0.0.1")
+            .put("http.host", "127.0.0.1")
+            .put("http.cors.enabled", true)
+            .put("http.cors.allow-origin", "*")
+            .put("auth.host_based.enabled", true)
+            .put("auth.host_based.config",
+                "a", new String[]{"user", "method", "protocol"}, new String[]{"John", "jwt", "http"})
+            .put("auth.host_based.config",
+                "b", new String[]{"user", "method"}, new String[]{"John", "password"})
+            .build();
+    }
+
+    /**
+     * Generates a random port (which affects url --> affects token's signature).
+     * Prepares a jwk endpoint response with pre-generated public key.
+     * Public key format is aligned with
+     * https://console.cratedb-dev.cloud/api/v2/meta/jwk/
+     * https://login.microsoftonline.com/common/discovery/v2.0/keys
+     * https://www.googleapis.com/oauth2/v3/certs
+     * and looks like:
+     * {
+     *   "keys":[
+     *     {
+     *       "e":"...",
+     *       "kid":"...",
+     *       "kty":"RSA",
+     *        "n":"..."
+     *     }
+     *    ]
+     * }
+     */
+    @Before
+    public void init() throws Exception {
+        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+        // Prepare private key.
+        PKCS8EncodedKeySpec privateKeySpec = new PKCS8EncodedKeySpec(BASE_64_DECODER.decode(PRIVATE_KEY_256));
+        privateKey = (RSAPrivateKey) keyFactory.generatePrivate(privateKeySpec);
+
+        // We use random port for the test suite.
+        // Port affects url --> affects signature --> need to re-compute payload.
+        port = RandomizedTest.randomIntBetween(10000, 20000);
```

Comment:
Usual pattern is to create a socket for port 0 to have the kernel assign a free port to avoid conflicts.
This can otherwise lead to hard to debug flaky tests.

Reply:
https://github.com/crate/crate/pull/15551/commits/1c235cb11111b2b0f0f5c851345423dcea2695ff